Hi Peter,

This is very cool! I experimented a while back trying to speed the code up with cython, but didn't really get anywhere.

I had a go at running the hizquasar example code with your numba branch, but I'm getting an error

`
Traceback (most recent call last):
  File "example_hizquasar.py", line 16, in <module>
    spec_resample, spec_errs_resample = spectres(regrid, spectrum[:,0], spectrum[:,1], spec_errs=spectrum[:,2])
  File "/Users/adam/anaconda/envs/py3/lib/python3.7/site-packages/numba/core/dispatcher.py", line 415, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/adam/anaconda/envs/py3/lib/python3.7/site-packages/numba/core/dispatcher.py", line 358, in error_rewrite
    reraise(type(e), e, None)
  File "/Users/adam/anaconda/envs/py3/lib/python3.7/site-packages/numba/core/utils.py", line 80, in reraise
    raise value.with_traceback(tb)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Can't unify return type from the following types: UniTuple(array(float64, 1d, C) x 2), array(float64, 1d, C)
Return of: IR name '$784return_value.1', type 'array(float64, 1d, C)', location: 
File "../spectres/spectral_resampling_numba.py", line 157:
def spectres(new_wavs, spec_wavs, spec_fluxes, spec_errs=None, fill=0,
    <source elided>
    else:
        return new_fluxes
        ^
Return of: IR name '$780return_value.3', type 'UniTuple(array(float64, 1d, C) x 2)', location: 
File "../spectres/spectral_resampling_numba.py", line 153:
def spectres(new_wavs, spec_wavs, spec_fluxes, spec_errs=None, fill=0,
    <source elided>
    if old_errs is not None:
        return new_fluxes, new_errs
        ^
`

--- a/spectres/spectral_resampling_numba.py
+++ b/spectres/spectral_resampling_numba.py
@@ -17,8 +17,24 @@ def make_bins(wavs):
 
     return edges, widths
 
-@jit(nopython=True, cache=True)
+
 def spectres(new_wavs, spec_wavs, spec_fluxes, spec_errs=None, fill=0,
+             verbose=True):
+    """ 
+    Wrapper function to preserve interface when using Numba
+    """
+
+    new_fluxes, new_errs = spnumba(new_wavs, spec_wavs, spec_fluxes, spec_errs=spec_errs, fill = fill, verbose=verbose)
+
+    #If errors were not supplied, we should only return the fluxes and not the array of zeros
+    if spec_errs is None:
+        return new_fluxes
+    #But if they were, we should return everything
+    else:
+        return new_fluxes, new_errs
+
+@jit(nopython=True, cache=True)
+def spnumba(new_wavs, spec_wavs, spec_fluxes, spec_errs=None, fill=0,
              verbose=True):
 
     """
@@ -84,8 +100,8 @@ def spectres(new_wavs, spec_wavs, spec_fluxes, spec_errs=None, fill=0,
         if old_errs.shape != old_fluxes.shape:
             raise ValueError("If specified, spec_errs must be the same shape "
                              "as spec_fluxes.")
-        else:
-            new_errs = np.copy(new_fluxes)
+    #    else:
+    new_errs = np.zeros_like(new_fluxes)
 
     start = 0
     stop = 0
@@ -148,10 +164,5 @@ def spectres(new_wavs, spec_wavs, spec_fluxes, spec_errs=None, fill=0,
             old_widths[start] /= start_factor
             old_widths[stop] /= end_factor
 
-    # If errors were supplied return both new_fluxes and new_errs.
-    if old_errs is not None:
-        return new_fluxes, new_errs
-
-    # Otherwise just return the new_fluxes spectrum array
-    else:
-        return new_fluxes
+    #numba doesn't seem to like having more than one return, so we're forcing it to return an array of zeros for the errors here
+    return new_fluxes, new_errs

--- a/spectres/spectres_numba_speed_tests.py
+++ b/spectres/spectres_numba_speed_tests.py
@@ -1,0 +1,65 @@
+import numpy as np
+from timeit import timeit, Timer
+
+from spectral_resampling_numba import spectres as spnumba
+from spectral_resampling import spectres as sp
+
+def call_spectres():
+    inwaves = np.linspace(0.5, 15, size)
+    outwaves = np.linspace(1, 10, 200)
+    influxes = np.ones_like(inwaves)
+    r = sp(outwaves, inwaves, influxes)
+    return r
+
+def call_spectres_err():
+    inwaves = np.linspace(0.5, 15, size)
+    outwaves = np.linspace(1, 10, 200)
+    influxes = np.ones_like(inwaves)
+    inerrs = 0.1*influxes
+    r = sp(outwaves, inwaves, influxes, spec_errs = inerrs)
+    return r
+
+
+def call_spectresn():
+    inwaves = np.linspace(0.5, 15, size)
+    outwaves = np.linspace(1, 10, 200)
+    influxes = np.ones_like(inwaves)
+    r = spnumba(outwaves, inwaves, influxes)
+    return r
+
+def call_spectresn_err():
+    inwaves = np.linspace(0.5, 15, size)
+    outwaves = np.linspace(1, 10, 200)
+    influxes = np.ones_like(inwaves)
+    inerrs = 0.1*influxes
+    r = spnumba(outwaves, inwaves, influxes, spec_errs = inerrs)
+    return r
+
+
+if __name__=="__main__":
+    insizes = [1000,10000, 100000]
+    repeats = 1000
+    size = insizes[0]
+    a = call_spectres_err()
+
+    b = call_spectresn_err()
+
+    for size in insizes:
+        
+        print("Running spectres with ",size," input wavelengths for ",repeats," times")
+        timer = Timer(stmt='call_spectres()', setup = 'from spectral_resampling import spectres as sp; from __main__ import call_spectres') 
+        time = timer.timeit(repeats)
+        print("Total runtime = ",time,"s, time per call = ",time/repeats,"s")
+
+        call_spectresn() #Call the compiled version once with the same argument types so that it's ready for the speed test - this could really be done outside the loop
+        print("Running compiled version of spectres with ",size," input wavelengths for ",repeats," times")
+        timer = Timer(stmt='call_spectresn()', setup = 'from spectral_resampling import spectres as sp; from __main__ import call_spectresn')
+        time_c = timer.timeit(repeats)
+
+        print("Total runtime = ",time_c,"s, time per call = ",time_c/repeats,"s")
+        print("Speedup thanks to Numba = ",time/time_c)
+
+        
+        
+
+    

